src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1335,14 +1335,20 @@ private BadHttpRequestException GetInvalidRequestTargetException(ReadOnlySpan<by
         public void SetBadRequestState(BadHttpRequestException ex)
         {
             Log.ConnectionBadRequest(ConnectionId, ex);
+            _requestRejectedException = ex;
 
             if (!HasResponseStarted)
             {
                 SetErrorResponseException(ex);
             }
 
+            const string badRequestEventName = "Microsoft.AspNetCore.Server.Kestrel.BadRequest";
+            if (ServiceContext.DiagnosticSource?.IsEnabled(badRequestEventName) == true)
+            {
+                ServiceContext.DiagnosticSource.Write(badRequestEventName, this);
```

Comment:
So what does this look like on the receivers end? Write takes an object so they'll need to downcast. What are they down casting to, IBadRequestExceptionFeature? If so, then how do they get the other request fields? Assuming you can downcast the same object to multiple interfaces seems like a usage bad pattern to encourage. Should we be telling them to always downcast to IFeatureCollection and the Get the features they need?

Reply:
> Should we be telling them to always downcast to IFeatureCollection and the Get the features they need?

Yes.

Reply:
Since the listener is going to have to downcast to `IFeatureCollection` anyway and they can now get the status code from the `IHttpResponseFeature`, I think it makes even more sense to just make `IBadRequestExceptionFeature.Error` a plain-old `Exception` and then we can put the feature back where it was.

Reply:
That makes sense. Will update.